### PR TITLE
[FSDP] Add set_state_dict_type API to setup state_dict_type without using context manager

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86243

FSDP.state_dict_type is a context manager. However, users may want to decide what state_dict is going to used during initialization. `set_state_dict_type` allows users to do so.

Differential Revision: [D40083670](https://our.internmc.facebook.com/intern/diff/D40083670/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2059,6 +2053,46 @@ def _get_training_state(
...
+    ) -> Tuple[StateDictType, StateDictConfig]:
+        _state_dict_type_to_config = {
+            StateDictType.FULL_STATE_DICT: FullStateDictConfig,
+            StateDictType.LOCAL_STATE_DICT: LocalStateDictConfig,
+            StateDictType.SHARDED_STATE_DICT: ShardedStateDictConfig,
+        }
+
+        prev_state_dict_type = None
+        prev_state_dict_config = None
+        # Use default config a state_dict config is not set.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2070]: "_if_ a state_dict config is not set"

**Review by rohan-varma — APPROVED** [review]
LGTM, just looks like there is a valid lint. Thanks for adding this!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -2111,6 +2104,46 @@ def _get_training_state(
         )
         return next(iter(training_states))
 
+    @staticmethod
+    def set_state_dict_type(
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2108]: Should we add a docstring?

**Pull request closed without merging**